src/Microsoft.AspNet.Routing/Template/TemplateMatcher.cs

```diff
@@ -40,126 +40,175 @@ public class TemplateMatcher
+                    break;
+                }
 
-                // It's ok for a catch-all to produce a null value
-                object defaultValue;
-                if (Defaults.TryGetValue(part.Name, out defaultValue) || part.IsCatchAll)
+                // If we get here, this is a simple segment with a parameter. We need it to be optional, or for the
+                // defaults to have a value.
+                Debug.Assert(routeSegment.IsSimple && part.IsParameter);
+                if (!Defaults.ContainsKey(part.Name) && !part.IsOptional)
                 {
-                    values.Add(part.Name, defaultValue);
+                    // There's no default for this (non-optional) parameter so it can't match.
+                    return null;
                 }
-                else if (part.IsOptional)
+            }
+
+            // At this point we've very likely got a match, so start capturing values for real.
+            var values = new RouteValueDictionary();
+
+            i = 0;
+            foreach (var requestSegment in pathTokenizer)
+            {
+                var routeSegment = Template.GetSegment(i++);
+
+                if (routeSegment.IsSimple && routeSegment.Parts[0].IsCatchAll)
                 {
-                    // This is optional (with no default value) - there's nothing to capture here, so just move on.
+                    // A catch-all captures til the end of the string.
+                    var part = routeSegment.Parts[0];
+                    var captured = requestSegment.GetRemainingPath();
+                    if (captured.Length > 0)
+                    {
+                        values.Add(part.Name, captured);
+                    }
+                    else
+                    {
+                        // It's ok for a catch-all to produce a null value
+                        object defaultValue;
+                        Defaults.TryGetValue(part.Name, out defaultValue);
+
+                        values.Add(part.Name, defaultValue);
+                    }
+
+                    // A catch-all has to be the last part, so we're done.
+                    break;
                 }
-                else
+                else if (routeSegment.IsSimple && routeSegment.Parts[0].IsParameter)
                 {
-                    // There's no default for this (non-catch-all) parameter so it can't match.
-                    return null;
+                    // A simple parameter captures the whole segment, or a default value if nothing was
+                    // provided.
+                    var part = routeSegment.Parts[0];
+                    if (requestSegment.Length > 0)
+                    {
+                        values.Add(part.Name, requestSegment.ToString());
+                    }
+                    else
+                    {
+                        object defaultValue;
+                        if (Defaults.TryGetValue(part.Name, out defaultValue))
```

Comment:
The case you're talking about is for a catch-all, which we handle separately in this code.